core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -1958,7 +1959,33 @@ private RelOptUtil.Exists convertExists(
       if (!values.getTuples().isEmpty()) {
         unionInputs.add(values);
       }
-      resultRel = LogicalUnion.create(unionInputs, true);
+      RelDataType finalRowType = rowType;
```

Comment:
This is a nice optimization, but I don't understand how it relates to the issue that is being solved.
This should probably be a separate PR.
Moreover, this looks like this should be part of UNION_TO_VALUES, and not always part of SqlToRelConverter.

Reply:
﻿@mihaibudiu Yes, as you can see, it is the same as UNION_TO_VALUES but works in different phases. One is the convert SQL to RelNode phase, and the other is the optimize phase.
﻿
Before this PR. When we convert the SQL to RelNode(convert IN list to Join):
```
SELECT empno FROM emp AS e WHERE cast(e.empno as bigint) in (130, 131, 132, 133, 134)
```
The generated RelNode contains:
﻿
```
LogicalAggregate(group=[{0}])
  LogicalValues(tuples=[[{ 130 }, { 131 }, { 132 }, { 133 }, { 134 }]])
```
But when we add CAST in the IN list value:
The generated RelNode will be:
```
LogicalUnion(all=[true])
  LogicalValues(tuples=[[{ 130 }]])
  LogicalValues(tuples=[[{ 131 }]])
  LogicalValues(tuples=[[{ 132 }]])
  LogicalValues(tuples=[[{ 133 }]])
  LogicalValues(tuples=[[{ 134 }]])
```
I want to keep this unchanged, so add the same processing method here.
The unit tests in `SqlToRelConverterTest#testInToSemiJoin`.

Reply:
I think you are fixing a symptom of the bug, and not the actual bug.
This fix should work no matter which subquery is present. It would work even if the subquery is a union, or a LogicalValues, or any other query.
The fact that you have to handle union of values specially suggests that you are not fixing the bug in the issue, but another bug. Maybe the issue is wrong, and this is a bug with Union.

Reply:
The way to find out is to write another test query, where the query in the IN right-hand-side returns a result with a type that will also require a cast.

Reply:
Are you referring to this SQL?
```
select * from dept where cast(deptno as bigint) in (select deptno + 20 from dept)
```
I add a test case in SqlToRelConverterTest#testInSubQueryWithTypeCast.

Reply:
whatever fixes this query should also handle the union case.

Reply:
The commit info describes a usage phenomenon caused by this defect, not an actual bug description.

Reply:
Where is the actual bug that is being fixed?

Reply:
> whatever fixes this query should also handle the union case.

Can you give a unit test case or an example SQL? I will cover it.

Reply:
In Calcite, we don't add cast if the source type precedence list contains a target type if type is integer types (smallint, tinyint, integer, and bigint). But if we don't add cast, the SQL output will be wrong. 
So In this PR, I will add cast even if the source type precedence list contains the target type(only works in integer types).  The corresponding code in AbstractTypeCoercion.java.

Reply:
Great, so you deleted code which was omitting the casts.
The question I have: is this change handling union in the SqlToRelConverter required to make the test pass? If it's still required, it means that there is somewhere another bug. 
If it's not required, perhaps you don't need to add this change about union in this PR.

Reply:
SqlToRelConverter is still required. Because it occurred during the conversion of IN list to Values.

Before this PR. 

The SQL:
```
SELECT empno FROM emp AS e WHERE cast(e.empno as bigint) in (130, 131, 132, 133, 134)
```
The Calcite will convert it to :
```
LogicalAggregate(group=[{0}])
  LogicalValues(tuples=[[{ 130 }, { 131 }, { 132 }, { 133 }, { 134 }]])
```
But when we add CAST like:
```
SELECT empno FROM emp AS e WHERE cast(e.empno as bigint) in (cast(130 as bigint), cast(131 as bigint), cast(132 as bigint), cast(133 as bigint), cast(134 as bigint))
````
The Calcite will convert it to:
```
LogicalUnion(all=[true])
  LogicalValues(tuples=[[{ 130 }]])
  LogicalValues(tuples=[[{ 131 }]])
  LogicalValues(tuples=[[{ 132 }]])
  LogicalValues(tuples=[[{ 133 }]])
  LogicalValues(tuples=[[{ 134 }]])
```
This test case was not covered before. The change in this PR makes this bug apparent. Do we need a separate PR to resolve it first? I hope this can answer your question.

Reply:
Why is the form with union not correct? Has the cast information been lost?

Reply:
The Union is correct, but VALUE is simpler. Not lost anything in UNION.
Before this PR, the IN-LIST SQL without cast can be converted to values, I hope it can still be done here after we add the CAST.

Reply:
If it works, I would prefer a solution in which the UNION rewrite is done using the UNION_TO_VALUES rule. There is no point in duplicating the logic of that rule in the SqlToRelConverter.

Reply:
Yes, I get it. UNION_TO_VALUES works in the optimization phase. But here is to convert the SQL to the RelNode phase. 
Before this PR. SQL like `cast(number as bigint) in (12,13,14) ` can convert to 'VALUES' when converting the SQL to the RelNode. I hope to keep the semantics here unchanged, so I added it here. If we don't do this, the RelNode converted here will be different from the previous version. According to the review comment in Jira https://issues.apache.org/jira/browse/CALCITE-5156 it is unacceptable.

Reply:
But if you leave out the optimization of union from SqlToRelConverter, won't the plans in RelOptRulesTest look more like they used to?

I think that the insertion of the cast is unavoidable, that's after all what this change is about.

Reply:
I understand what you mean, but I need to clarify that this does not include all UNION. It is only when converting IN-LIST to values. If you think we should remove the changes in SqlToRelConverter then I will do it. I think this is not a bug, as long as we reach an agreement, it's fine.

Reply:
I think SqlToRelConverter should do the minimal amount of rewrites necessary to implement the program with the correct semantics, and no optimizations. The code is already complex, it's best to separate the concerns. But, clearly, the Rel representation produced should contain enough information to represent the programs, including some casts that cannot be inferred otherwise.

Reply:
@mihaibudiu I have rollback the SqlToRelConverter. Now this PR includes CALCITE-6592. If you have free time, please review it.

Reply:
I have already approved these changes.